### PR TITLE
Support .env file to set environment variables

### PR DESCRIPTION
**Pull request by robert197**
This pull request enables a confortable way to set environment variables.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,13 +13,68 @@ Use [Composer](https://getcomposer.org/) to install the testing framework.
 
 ```bash
 $ composer require --dev nimut/testing-framework
+$ composer require typo3/cms-core
+$ composer require typo3/cms
```
Comment (IchHabRecht) [inline README.md]: This is wrong and does not belong to the testing framework

File: README.md [hunk 2]
```diff
@@ -13,13 +13,68 @@ Use [Composer](https://getcomposer.org/) to install the testing framework.
...
 Composer will add the package as a dev requirement to your composer.json and install PHPUnit and vfsStream as its
 dependencies.
 
+or just edit your compose.json file:
+
+```json
+...
+"require": {
+    "typo3/cms-core": ">=7.6.0,<7.6.99",
+    "typo3/cms": "^7.6"
```
Comment (IchHabRecht) [inline README.md]: Wrong, the testing framework works with TYPO3 CMS 6.2 - 8.7

File: README.md [hunk 3]
```diff
@@ -13,13 +13,68 @@ Use [Composer](https://getcomposer.org/) to install the testing framework.
...
+  },
+  "require-dev": {
+    "nimut/testing-framework": "^1.0"
+  },
+...
+```
+And run ```composer install```
```
Comment (IchHabRecht) [inline README.md]: Wrong. If a composer.lock file exists an update is needed. Better use composer require as mentioned above.

File: README.md [hunk 4]
```diff
@@ -13,13 +13,68 @@ Use [Composer](https://getcomposer.org/) to install the testing framework.
...
+typo3DatabasePassword=your_test_db_password
+```
+
+Make sure the following things are defined in your `composer.json` file:
+
+```json
+      ...
+      "require-dev": {
+        "phpunit/phpunit": "~4.8.0",
+        "nimut/testing-framework": "^1.0"
```
Comment (IchHabRecht) [inline README.md]: phpunit is already required by the testing framework.

File: README.md [hunk 5]
```diff
@@ -13,13 +13,68 @@ Use [Composer](https://getcomposer.org/) to install the testing framework.
...
+      },
+      "autoload-dev": {
+        "psr-4": {
+          "My\\ExampleExtension\\Tests\\": "Tests"
```
Comment (IchHabRecht) [inline README.md]: Missing forward slash at the end of the folder name

File: README.md [hunk 6]
```diff
@@ -73,9 +128,12 @@ The nimut/testing-framework ships database fixtures for several TYPO3 CMS core d
 - sys_language
 - tt_content
 
-To use the database fixtures you can trigger an import in your test file
+Override `$testExtensionsToLoad` in your test class to create extension tables in the test database,
+and trigger an import:
```
Comment (IchHabRecht) [inline README.md]: Why did you remove the line above? This is incorrect and not what is explained with

$this->importDataSet('ntf://Database/pages.xml');

**Review by IchHabRecht — COMMENTED** [review]
I tend to decline this patch. Waiting for helhums comments.

**Comment by helhum** [comment]
@robert197 can you elaborate on your use case? Would using https://github.com/helhum/dotenv-connector cover your use case?

**Comment by helhum** [comment]
And you can specify env vars to be set in a phpunit.xml …

**Comment by helhum** [comment]
@robert197 I still don't understand your use case. 
What exactly is not covered with helhum/dotenv-connector?

File: README.md (hunk 1 above)
Comment (robert197) [reply README.md]: I thought it is a requirement because this testing framework is not working without typo3 core.
Also it wasn't documented.

File: README.md (hunk 3 above)
Comment (robert197) [reply README.md]: By versioning our composer.json, we want to run single command to install all dependencies. This way is usual in other technologies.

File: README.md (hunk 1 above)
Comment (IchHabRecht) [reply README.md]: The reasons for not mentioning those dependencies are they don't belong directly to the usage of the testing-framework.

typo3/cms-core should be required by the extension even without the testing-framework

typo3/cms is no dependency of the extension - at least only for develop context. but in aspect of the testing-framework this dependency is added by your CI server to be flexible.

Both commands only require the latest, stable version of TYPO3 CMS. As the testing-framework is designed to test against multiple versions at the same time, I would not recommend to add those to the documentation.

**Review by IchHabRecht — CHANGES_REQUESTED** [review]
Please answer the questions from Helmut (helhum). We still don't see any advantage in this pull request.

File: README.md (hunk 6 above)
Comment (robert197) [reply README.md]: While trying to load data fixtures it was not enaugh to to add ``$this->importDataSet('ntf://Database/pages.xml');``, we noticed that variable `$testExtensionsToLoad` was required too to make it work.

File: README.md (hunk 3 above)
Comment (IchHabRecht) [reply README.md]: Feel free to do so. However, I would recommend to add typo3/cms as a dev-dependency. And as I mentioned before, after you edit the composer.json file you need to run composer update instead of composer install.

**Comment by robert197** [comment]
@helhum We wanted to use a standard way, which is followed by other technologies in web development.
The reason and use case of this PR is to make tests work not only in CI but also locally after clonning an extension. We think tests should be ran localy before they are pushed into build server. At least this is the recommended way in software engineering. helhum/dotenv-connector is nice :+1: . but isnt it easier to put only .env file into the root of a project and make it work without doing extra stuff? I think this is how other frameworks work. I hope we didn't misunderstood this framework but if so, that would be great, if you could inform me :).

File: README.md (hunk 6 above)
Comment (IchHabRecht) [reply README.md]: This is independent from a database import. $testExtensionsToLoad is required to execute your functional tests anyways. Otherwise your extension isn't accessible for the tests.

See: https://wiki.typo3.org/Functional_testing#How_to_write_functional_test.3F and https://de.slideshare.net/cpsitgmbh/functional-tests-with-typo3

**Comment by helhum** [comment]
@robert197 if you have the testing framework as dev dependency, you can just add dotenv connector as dev dependency as well. then you can just put a .env file next to the composer.json and everything there is exposed as env vars to the testing framework without touching a line of code

File: README.md (hunk 1 above)
Comment (bittner) [reply README.md]: > should be required by the extension even without the testing-framework

I agree that a dependency that comes along with the unit under test, in any case, does not need to be explicitly listed. (In other frameworks unnecessarily listing dependencies has proven to cause more issues than advantages, IIRC.)

Comment (bittner) [reply README.md]: > this dependency is added by your CI server to be flexible.

You do assume that tests are run exclusively by the CI server? Let's agree that this is bad practice. We shouldn't presume such a setup. Tests should always be run locally first, as an engineers duty to "not ship broken code".

The CI server is meant to be an engineers colleague that doesn't forget to run _all_ the tests. We, as humans, can forget or be negligent. We may run a subset of the tests only, or be otherwise lazy. Disciplined engineers will try to avoid the latter.

File: README.md (hunk 6 above)
Comment (bittner) [reply README.md]: The instructions in a README are meant to be self-contained. Given we agree on that, can we list all essential details in it? A sequence of instructions should allow to reproduce the expected result. Assumptions should be mentioned clearly, right before the start of the sequence.

Where should the paragraph mentioning to "override `$testExtensionsToLoad` in your test class" go instead?

Comment (bittner) [reply README.md]: Oh, just noticed that this has been discussed in the past in issue #28. PR #29, upon your request, is waiting to be merged since then. Please, go ahead if it's fine. Thank you! :+1: 

(Though, it doesn't mention "override `$testExtensionsToLoad` in your test class" yet...)

Comment (IchHabRecht) [reply README.md]: The framework is meant to run your extension tests against multiple versions. It is based on the functionality of TYPO3 CMS 6.2 and 7.6 that was extracted to the typo3/testing-framework. Unfortunately with that extraction it isn't possible to test against multiple versions. It is not scope of that framework to explain how testing in extension with TYPO3 CMS works in general. We cannot offer those documentation and ask everyone to look in the already provided documentations of the community (see #43 ). As the $testExtensionsToLoad property is an important property to be able to run your tests with any of the both testing-frameworks, it is not related to nimut/testing-framework and not mentioned in this documentation. Otherwise we need to mention all important properties (which we cannot offer).

I really hope you understand that the functionality of the test cases and the concept of testing offered by this framework are extracted from already existing sources. And we cannot provide any support within the framework itself.

File: README.md (hunk 1 above)
Comment (IchHabRecht) [reply README.md]: I don't asume anything. I run tests locally on my own. But then as part of another project or with "composer require typo3/cms=^x.y" without committing the dependency. And currently I don't have any problems using the extension/framework like this. Maybe you find this blog post http://insight.helhum.io/post/148886148725/composerjson-specification-for-typo3-extensions interesting. There you find a recommendation of how your extensions composer.json should look like.

By the way: only require typo3/cms doesn't help you to run your tests as your extension within this typo3 installation is still missing. You have to create a link or copy your extension files to typo3conf/ext at the end. Either manually or by using some scripts.

**Comment by robert197** [comment]
ok, i tried to integrate helhum/dotenv-connector but because of php version its not possible. maybe that could be the reason to use the previous solution?

**Comment by helhum** [comment]
@robert197 helhum/dotenv-connector 1.x is compatible with PHP 5.3.7 or higher.

**Comment by robert197** [comment]
hey, is it possible to merge this changes?

**Comment by IchHabRecht** [comment]
Sorry, but for me it is still unclear why this change is needed and you cannot use helhum/dotenv-connector in your project?

**Comment by robert197** [comment]
@IchHabRecht i thought because of https://de.wikipedia.org/wiki/Don%E2%80%99t_repeat_yourself

We have a lot of extensions which are used to be maintained. It would be better for us to centralize some of dependencies.
Also less code or configuration is made in a SUT, less can be broken.

**Comment by helhum** [comment]
Thanks for your efforts.

However I still fail to see how we can improve things for our users without forcing them to use one solution (.env)

There are other ways to achieve the same like exporting the env vars in the shell or creating your own phpunit xml config and define env vars there.

You also can use the dotenv connector as pointed out in your setups. No change is required here for any of these solutions.

>  It would be better for us to centralize some of dependencies.

That is also easily possible. Just create your own testing package that requires everything you need
and only require this package within your extensions. 

Centralizing dependencies would mean forcing users to fetch a package which they don't need because the are more comfortable with other solution. The beauty of composer is giving control to users composing packages that fit their own needs.

Long story short: I'm closing this PR now. Everything you need as described in your use case is there. Just use it :)

**Comment by robert197** [comment]
but... this feature was optional...

**Comment by helhum** [comment]
a requirement is not optional ;)

**Comment by robert197** [comment]
okay nevermind, have a great day 👍

**Comment by helhum** [comment]
Thanks :) sorry for the back and forth and the misunderstandings!
